src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -97,37 +153,86 @@ private bool ApplicableOperatorIs(MethodSymbol candidate, CSharpSyntaxNode node,
             DiagnosticBag diagnostics,
             out bool wasExpression,
             bool wasSwitchCase)
         {
-            var expression = BindValue(right, diagnostics, BindValueKind.RValue);
+            var expression = BindValue(patternExpression, diagnostics, BindValueKind.RValue);
+            ConstantValue constantValueOpt = null;
+            expression = ConvertPatternExpression(operandType, patternExpression, expression, ref constantValueOpt, diagnostics);
             wasExpression = expression.Type?.IsErrorType() != true;
-            if (!node.HasErrors && expression.ConstantValue == null)
+            if (!expression.HasErrors && constantValueOpt == null)
             {
-                diagnostics.Add(ErrorCode.ERR_ConstantExpected, right.Location);
+                diagnostics.Add(ErrorCode.ERR_ConstantExpected, patternExpression.Location);
                 hasErrors = true;
             }
 
-            return new BoundConstantPattern(node, expression, hasErrors);
+            bool? knownMatchResult = null;
+            if (constantValueOpt != null)
+            {
+                if (constantValueOpt.IsNull)
+                {
+                    knownMatchResult = operandIsNull;
+                }
+                else if (operand.ConstantValue != null)
+                {
+                    knownMatchResult = constantValueOpt.Equals(operand.ConstantValue);
+                }
+            }
+
+            return new BoundConstantPattern(node, expression, constantValueOpt, knownMatchResult, hasErrors);
+        }
+
+        internal BoundExpression ConvertPatternExpression(TypeSymbol leftType, CSharpSyntaxNode node, BoundExpression expression, ref ConstantValue constantValue, DiagnosticBag diagnostics)
+        {
+            // NOTE: This will allow user-defined conversions, even though they're not allowed here.  This is acceptable
+            // because the result of a user-defined conversion does not have a ConstantValue and we'll report a diagnostic
+            // to that effect later.
+            BoundExpression convertedCaseExpression = GenerateConversionForAssignment(leftType, expression, diagnostics);
+
+            if (convertedCaseExpression.Kind == BoundKind.Conversion)
+            {
+                var conversion = (BoundConversion)convertedCaseExpression;
+                var operand = conversion.Operand;
+                if (leftType.IsNullableType() && (convertedCaseExpression.ConstantValue == null || !convertedCaseExpression.ConstantValue.IsNull))
+                {
+                    // Null is a special case here because we want to compare null to the Nullable<T> itself, not to the underlying type.
+                    var discardedDiagnostics = DiagnosticBag.GetInstance(); // We are not intested in the diagnostic that get created here
+                    convertedCaseExpression = CreateConversion(operand, leftType.GetNullableUnderlyingType(), discardedDiagnostics);
```

Comment:
That would only do extra work to produce diagnostics for assignment that we are not interested in.

---
In reply to: [64944989](https://github.com/dotnet/roslyn/pull/11485#discussion_r64944989) [](ancestors = 64944989)